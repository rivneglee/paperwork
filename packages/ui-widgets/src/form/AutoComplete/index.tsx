export { default as AutoComplete } from './AutoComplete';
export * from './AutoComplete';
