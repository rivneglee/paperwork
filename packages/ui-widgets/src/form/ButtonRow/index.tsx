export { default as ButtonRow } from './ButtonRow';
