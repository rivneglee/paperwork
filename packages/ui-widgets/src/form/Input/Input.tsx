import React, { ReactElement, SyntheticEvent } from 'react';
import Cleave from 'cleave.js/react';
import classNames from 'classnames';

import { FieldGroup } from '../FieldGroup';

interface Props {
  value?: string | number;
  label?: ReactElement | string;
  disabled?: boolean;
  readOnly?: boolean;
  isRequired?: boolean;
  labelAccessory?: ReactElement;
  labelPlacement?: 'left' | 'top';
  onChange?: (e: SyntheticEvent) => void;
  onKeyDown?: (e: SyntheticEvent) => void;
  size?: 'xs' | 's' | 'm' | 'l' | 'xl';
  placeholder?: string;
  options?: object;
  type?: 'outlined' | 'underlined';
  left?: ReactElement;
  right?: ReactElement;
  className?: string;
}

class Input extends React.Component<Props> {
  state = {
    hasFocus: false,
  };

  private onInputFocus = () => this.setState({
    hasFocus: true,
  })

  private onInputBlur = () => this.setState({
    hasFocus: false,
  })

  render() {
    const {
      type = 'outlined',
      label,
      isRequired,
      labelAccessory,
      labelPlacement,
      options,
      size,
      left,
      right,
      className,
      value = '',
      ...otherProps
    } = this.props;

    const { hasFocus } = this.state;
    const inputProps = {
      value,
      ...otherProps,
      className: 'pw-input__control',
      onFocus: this.onInputFocus,
      onBlur: this.onInputBlur,
    };

    const inputView = options ? (
      <Cleave
        {...inputProps}
        options={options}
      />
    ) : <input {...inputProps} />;

    return (
      <FieldGroup
        label={label}
        isRequired={isRequired}
        labelAccessory={labelAccessory}
        size={size}
        labelPlacement={labelPlacement}
      >
        <div
          className={
            classNames(
              'pw-input',
              hasFocus && 'pw-input--active',
              type && `pw-input--type-${type}`,
              className,
            )
          }>
          {
            left && <span className="pw-input__left">{left}</span>
          }
          {inputView}
          {
            right && <span className="pw-input__right">{right}</span>
          }
        </div>
      </FieldGroup>
    );
  }
}

export default Input;
