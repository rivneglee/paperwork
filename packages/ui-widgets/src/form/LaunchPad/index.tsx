export { default as LaunchPad } from './LaunchPad';
