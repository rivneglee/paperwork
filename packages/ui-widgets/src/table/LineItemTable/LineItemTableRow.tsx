import React, { FunctionComponent, ReactElement } from 'react';
import { Table } from '../Table';

export interface LineItemTableColumnConfig {
  columnName: string;
  className?: string;
  hideLabel?: boolean;
}

interface Props {
  columnsConfig: LineItemTableColumnConfig[];
}

export interface LineItemTableRowComponent extends FunctionComponent<Props> {}

const LineItemTableRow: LineItemTableRowComponent = ({ children, columnsConfig }) => {
  const columns = children instanceof Array ? children : [children];
  return (
    <>
      {
        columns.map((columnElement: ReactElement, index) => (
          <Table.RowItem
            key={columnsConfig[index].columnName}
            {...columnsConfig[index]}
          >
            {columnElement}
          </Table.RowItem>
        ))
      }
    </>
  );
};

export default LineItemTableRow;
