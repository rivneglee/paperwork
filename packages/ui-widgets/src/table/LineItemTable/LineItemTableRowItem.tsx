import React, { FunctionComponent } from 'react';

interface Props {}

export interface LineItemTableRowItemComponent extends FunctionComponent<Props> {}

const LineItemTableRowItem: LineItemTableRowItemComponent = ({ children }) => {
  return (
    <>
      {children}
    </>
  );
}

export default LineItemTableRowItem;
