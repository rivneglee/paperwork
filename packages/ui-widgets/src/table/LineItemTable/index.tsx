export { default as LineItemTable } from './LineItemTable';
