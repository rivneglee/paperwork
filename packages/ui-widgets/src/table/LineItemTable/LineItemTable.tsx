import React, { FunctionComponent, ReactElement } from 'react';
import shortid from 'shortid';

import { Table } from '../Table';
import Icons from '../../graphic/Icons';

import LineItemTableRow, { LineItemTableRowComponent, LineItemTableColumnConfig } from './LineItemTableRow';
import LineItemTableRowItem, { LineItemTableRowItemComponent } from './LineItemTableRowItem';
import { IconButton } from '../../form/IconButton';

interface Props {
  data: any[];
  columnsConfig: LineItemTableColumnConfig[];
  renderRow: (index: number, rowData: any, onUpdateRow: (...args: any) => void) => ReactElement;
  onUpdateRow?: (index: number, ...args: any) => void;
  onAddRow?: (...args: any) => void;
  onRemoveRow?: (index: number) => void;
}

interface LineItemTableComponent extends FunctionComponent<Props> {
  Row: LineItemTableRowComponent;
  Item: LineItemTableRowItemComponent;
}

const LineItemTable: LineItemTableComponent = ({
 children,
 data,
 columnsConfig,
 renderRow,
 onUpdateRow,
 onAddRow,
 onRemoveRow,
}) => {
  data.forEach((row: any) => {
    if (!row.id) {
      row.id = shortid.generate();
    }
  });

  const newRow = { id: shortid.generate() };

  const rows = [...data, newRow];

  const onChange = (index: number, rowData: any, ...args: any) => {
    if (index === data.length) {
      onAddRow && onAddRow(rowData, ...args);
    } else {
      onUpdateRow && onUpdateRow(index, ...args);
    }
  };

  return (
    <Table>
      <Table.Header>
        {
          columnsConfig.map(({ columnName, ...otherProps }) => (
            <Table.HeaderItem
              key={columnName}
              {...otherProps}
            >
              {columnName}
            </Table.HeaderItem>
          ))
        }
        <Table.RowItem className="pw-line-item-table__remove"/>
      </Table.Header>
      <Table.Body>
        {
          rows.map((rowData, index) => (
            <Table.Row key={rowData.id} className="pw-line-item-table__row">
              {renderRow(index, rowData, (...args) => onChange(index, rowData, ...args))}
              <Table.RowItem className="pw-line-item-table__remove" columnName="Remove" hideLabel>
                <IconButton
                  className="pw-line-item-table__remove-button"
                  onClick={() => onRemoveRow && onRemoveRow(index)}
                >
                  <Icons.Trash />
                </IconButton>
              </Table.RowItem>
            </Table.Row>
          ))
        }
      </Table.Body>
    </Table>
  );
};

LineItemTable.Row = LineItemTableRow;
LineItemTable.Item = LineItemTableRowItem;

export default LineItemTable;
