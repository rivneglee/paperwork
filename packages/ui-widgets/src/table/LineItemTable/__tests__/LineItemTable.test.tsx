import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import LineItemTable from '../LineItemTable';
import { Input } from '../../../form/Input';

describe('LineItemTable', () => {
  let wrapper: ShallowWrapper;

  const data = [
    { name: 'CHENG LI', team: 'Customer Service', date: '02/03/2020' },
    { name: 'SUN WEI', team: 'Customer Service', date: '02/03/2020' },
    { name: 'LIPING ZHAO', team: 'Customer Service', date: '02/03/2020' },
  ];

  const columnsConfig = [
    { columnName: 'Name' },
    { columnName: 'Team' },
    { columnName: 'On boarding date', style: { width: 200 } },
  ];

  beforeAll(() => {
    wrapper = shallow(
      <LineItemTable
        columnsConfig={columnsConfig}
        data={data}
        renderRow={(index, data) => (
          <LineItemTable.Row columnsConfig={columnsConfig}>
            <LineItemTable.Item>
              <Input value={data.name}/>
            </LineItemTable.Item>
            <LineItemTable.Item>
              <Input value={data.team}/>
            </LineItemTable.Item>
            <LineItemTable.Item>
              <Input value={data.date}/>
            </LineItemTable.Item>
          </LineItemTable.Row>
        )}
      />,
    );
  });

  it('should render properly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
