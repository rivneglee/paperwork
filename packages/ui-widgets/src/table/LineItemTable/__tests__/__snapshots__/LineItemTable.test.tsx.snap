// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineItemTable should render properly 1`] = `
<Table>
  <Header>
    <HeaderItem
      key="Name"
    >
      Name
    </HeaderItem>
    <HeaderItem
      key="Team"
    >
      Team
    </HeaderItem>
    <HeaderItem
      key="On boarding date"
      style={
        Object {
          "width": 200,
        }
      }
    >
      On boarding date
    </HeaderItem>
    <RowItem
      className="pw-line-item-table__remove"
    />
  </Header>
  <Body>
    <Row
      className="pw-line-item-table__row"
      key="1"
    >
      <LineItemTableRow
        columnsConfig={
          Array [
            Object {
              "columnName": "Name",
            },
            Object {
              "columnName": "Team",
            },
            Object {
              "columnName": "On boarding date",
              "style": Object {
                "width": 200,
              },
            },
          ]
        }
      >
        <LineItemTableRowItem>
          <Input
            value="CHENG LI"
          />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input
            value="Customer Service"
          />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input
            value="02/03/2020"
          />
        </LineItemTableRowItem>
      </LineItemTableRow>
      <RowItem
        className="pw-line-item-table__remove"
        columnName="Remove"
        hideLabel={true}
      >
        <IconButton
          className="pw-line-item-table__remove-button"
          onClick={[Function]}
        >
          <Component />
        </IconButton>
      </RowItem>
    </Row>
    <Row
      className="pw-line-item-table__row"
      key="2"
    >
      <LineItemTableRow
        columnsConfig={
          Array [
            Object {
              "columnName": "Name",
            },
            Object {
              "columnName": "Team",
            },
            Object {
              "columnName": "On boarding date",
              "style": Object {
                "width": 200,
              },
            },
          ]
        }
      >
        <LineItemTableRowItem>
          <Input
            value="SUN WEI"
          />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input
            value="Customer Service"
          />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input
            value="02/03/2020"
          />
        </LineItemTableRowItem>
      </LineItemTableRow>
      <RowItem
        className="pw-line-item-table__remove"
        columnName="Remove"
        hideLabel={true}
      >
        <IconButton
          className="pw-line-item-table__remove-button"
          onClick={[Function]}
        >
          <Component />
        </IconButton>
      </RowItem>
    </Row>
    <Row
      className="pw-line-item-table__row"
      key="3"
    >
      <LineItemTableRow
        columnsConfig={
          Array [
            Object {
              "columnName": "Name",
            },
            Object {
              "columnName": "Team",
            },
            Object {
              "columnName": "On boarding date",
              "style": Object {
                "width": 200,
              },
            },
          ]
        }
      >
        <LineItemTableRowItem>
          <Input
            value="LIPING ZHAO"
          />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input
            value="Customer Service"
          />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input
            value="02/03/2020"
          />
        </LineItemTableRowItem>
      </LineItemTableRow>
      <RowItem
        className="pw-line-item-table__remove"
        columnName="Remove"
        hideLabel={true}
      >
        <IconButton
          className="pw-line-item-table__remove-button"
          onClick={[Function]}
        >
          <Component />
        </IconButton>
      </RowItem>
    </Row>
    <Row
      className="pw-line-item-table__row"
      key="123"
    >
      <LineItemTableRow
        columnsConfig={
          Array [
            Object {
              "columnName": "Name",
            },
            Object {
              "columnName": "Team",
            },
            Object {
              "columnName": "On boarding date",
              "style": Object {
                "width": 200,
              },
            },
          ]
        }
      >
        <LineItemTableRowItem>
          <Input />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input />
        </LineItemTableRowItem>
        <LineItemTableRowItem>
          <Input />
        </LineItemTableRowItem>
      </LineItemTableRow>
      <RowItem
        className="pw-line-item-table__remove"
        columnName="Remove"
        hideLabel={true}
      >
        <IconButton
          className="pw-line-item-table__remove-button"
          onClick={[Function]}
        >
          <Component />
        </IconButton>
      </RowItem>
    </Row>
  </Body>
</Table>
`;
